File: src/main/java/freezingoctosansa/BinaryHeap.java
package freezingoctosansa;


public class BinaryHeap {

  public int _capacity;
  public Integer[] _arr;
  public int _size = 0;

  public BinaryHeap(int h) {
    _capacity = h;
    _arr = new Integer[(int) (Math.pow(2, _capacity) - 1)];
  }

  public int size() {
    return _size;
  }

  public void insert(int x) {
    if (_size == _arr.length) {
      throw new RuntimeException("heap is full");
    }

    _size++;
    _arr[_size - 1] = x;
    siftUp(_size - 1);
  }

  private void siftUp(int i) {
    if (i == 0) return;

    int parentIdx = getParent(i);
    if (_arr[parentIdx] > _arr[i]) {
      int tmp = _arr[parentIdx];
      _arr[parentIdx] = _arr[i];
      _arr[i] = tmp;
      siftUp(parentIdx);
    }
  }

  public int deleteMin() {
    if (_size == 0) throw new RuntimeException("heap is empty");
    int min = _arr[0];
    _arr[0] = _arr[_size - 1];
    _size--;
    siftDown(0);
    return min;
  }

  private void siftDown(int i) {
    int leftIdx = getLeft(i);
    int rightIdx = getRight(i);
    int minIdx;

    if (rightIdx >= _size) {
      if (leftIdx >= _size) {
        return;
      } else {
        minIdx = leftIdx;
      }
    } else {
      if (_arr[leftIdx] <= _arr[rightIdx]) {
        minIdx = leftIdx;
      } else {
        minIdx = rightIdx;
      }
    }

    if (_arr[i] > _arr[minIdx]) {
      int tmp = _arr[minIdx];
      _arr[minIdx] = _arr[i];
      _arr[i] = tmp;
      siftDown(minIdx);
    }
  }

  public void merge(BinaryHeap heap) {
    while(heap.size() > 0) {
      insert(heap.deleteMin());
    }
  }

  public void print() {
    print(0, 0, "*");
  }

  private void swap(int i, int j)
  {
    int tmp = _arr[j];
    _arr[j] = _arr[i];
    _arr[i] = tmp;
  }

  private void print(int i, int level, String prefix) {
    if (i >= _size) return;
    for (int j = 0; j < level; j++) System.out.print(" ");
    System.out.println(prefix + _arr[i]);
    print(getLeft(i), level + 1, "<");
    print(getRight(i), level + 1, ">");
  }

  private int getLeft(int i) {
    return (2 * i + 1);
  }

  private int getRight(int i) {
    return (2 * i + 2);
  }

  private int getParent(int i) {
    return ((i-1)/2);
  }
}

File: python/binary_heap.py
import math

class BinaryHeap:

    def __init__(self, h):
        self.h = h
        self.arr = [0]*int(math.pow(2,h) - 1)
        self.size = 0

    def insert(self, x):
        if self.size == len(self.arr):
            raise RuntimeError("heap is full")

        self.arr[self.size] = x
        self.size += 1
        self.sift_up(self.size - 1)

    @staticmethod
    def get_left(i):
        return 2*i + 1

    @staticmethod
    def get_right(i):
        return 2*i + 2

    @staticmethod
    def get_parent(i):
        return (i-1)//2

    def swap(self, i, j):
        self.arr[i], self.arr[j] = self.arr[j], self.arr[i]

    def sift_up(self, i):
        if i == 0:
            return

        parent_idx = self.get_parent(i)
        if self.arr[parent_idx] > self.arr[i]:
            self.swap(i, parent_idx)
            self.sift_up(parent_idx)

    def delete_min(self):
        if self.size == 0:
            raise RuntimeError("heap is empty")

        m = self.arr[0]
        self.arr[0] = self.arr[self.size - 1]
        self.size -= 1
        self.sift_down(0)
        return m

    def sift_down(self, i):
        left_idx = self.get_left(i)
        right_idx = self.get_right(i)

        min_idx = i
        if left_idx < self.size and self.arr[left_idx] < self.arr[i]:
            min_idx = left_idx
        if right_idx < self.size and self.arr[right_idx] < self.arr[i]:
            min_idx = right_idx

        if min_idx != i:
            self.swap(i, min_idx)
            self.sift_down(min_idx)

    def heapify(self):
        for i in range(self.size//2,-1,-1):
            self.sift_down(i)

    def merge(self, heap):
        self.h += heap.h
        new_arr = [0]*int(math.pow(2,self.h) - 1)
        for i in range(self.size):
            new_arr[i] = self.arr[i]
        for i in range(heap.size):
            new_arr[i+self.size] = heap.arr[i]
        self.arr = new_arr
        self.size += heap.size
        self.heapify()

    def print(self, i=0, level=0, prefix="*"):
        if i >= self.size:
            return
        print(" "*level, end="")
        print(prefix + str(self.arr[i]))
        self.print(self.get_left(i), level + 1, prefix="<")
        self.print(self.get_right(i), level + 1, prefix=">")

if __name__ == "__main__":
    import random
    random.seed(0)
    bh = BinaryHeap(10)
    bh2 = BinaryHeap(10)
    bh.print()
    for i in range(5):
      bh.insert(random.randint(0,100))
      bh2.insert(random.randint(0,100))
      bh.print()
      print("__________")
    bh.print()
    print("done")

    print("__________")
    print("delete min")
    bh.delete_min()
    bh.print()

    print("__________")
    print("merge")
    print("bh:")
    bh.print()
    print("bh2:")
    bh2.print()
    bh.merge(bh2)
    print("final:")
    bh.print()

